components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/basicauth/BasicAuthCredentialValidator.java

```
@@ -0,0 +1,478 @@
+            return authType;
+        }
+        return APIConstants.NO_MATCHING_AUTH_SCHEME;
+    }
+
+    /**
+     * Validates the given username and password against the users in the user store.
+     *
+     * @param username given username
+     * @param password given password
+     * @return true if the validation passed
+     * @throws APISecurityException If an authentication failure or some other error occurs
+     */
+    @MethodStats
+    public boolean validate(String username, String password) throws APISecurityException {
+        String providedPasswordHash = null;
+        if (gatewayKeyCacheEnabled) {
+            providedPasswordHash = hashString(password);
+            String cachedPasswordHash = (String) getGatewayUsernameCache().get(username);
+            if (cachedPasswordHash != null && cachedPasswordHash.equals(providedPasswordHash)) {
+                log.debug("Basic Authentication: <Valid Username Cache> Username & password authenticated");
+                return true; //If (username->password) is in the valid cache
+            } else {
+                String invalidCachedPasswordHash = (String) getInvalidUsernameCache().get(username);
+                if (invalidCachedPasswordHash != null && invalidCachedPasswordHash.equals(providedPasswordHash)) {
+                    log.debug("Basic Authentication: <Invalid Username Cache> Username & password authentication failed");
+                    return false; //If (username->password) is in the invalid cache
+                }
+            }
+        }
+
+        boolean authenticated;
+        try {
+            authenticated = authAdminStub.login(username, password, host);
+        } catch (RemoteException | LoginAuthenticationExceptionException e) {
+            log.debug("Basic Authentication: Username and Password authentication failure");
+            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR, e.getMessage());
```

Exception has been swallowed leading to the original stack trace being lost. Pass entire exception object into wrapped exception

catch (CaughtExecption e) { throw new ThrownException("Error detected", e); }

---

Updated